Code generation documentation

tokencolour opened the pull request:
Added changes as suggested by @asmeurer and @moorepants.
Not yet changed the residence of the file.
Extends #10276
Fixes #8889

tokencolour committed:
Code generation documentation

Added changes as suggested by @asmeurer and @moorepants.
Not yet changed the residence of the file.
Extends #10276
Fixes #8889

moorepants commented:
These look like pretty printing errors:

```
File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 65, in codegen.rst

Failed example:

    Assignment(known_mat, mat)

Expected:

    K := [x  y  z]

Got:

    Assignment(K, Matrix([[x, y, z]]))

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 69, in codegen.rst

Failed example:

    Assignment(known_mat, mat).rhs

Expected:

    [x  y  z]

Got:

    Matrix([[x, y, z]])

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 77, in codegen.rst

Failed example:

    expr

Expected:

        2   

    -Z⋅e ⋅k 

    ────────

      2⋅r   

Got:

    -Z*e**2*k/(2*r)

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 82, in codegen.rst

Failed example:

    ccode(expr)

Expected:

    -1.0L/2.0L*Z*pow(e, 2)*k/r

Got:

    '-1.0L/2.0L*Z*pow(e, 2)*k/r'

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 84, in codegen.rst

Failed example:

    ccode(expr, assign_to="E")

Expected:

    E = -1.0L/2.0L*Z*pow(e, 2)*k/r;

Got:

    'E = -1.0L/2.0L*Z*pow(e, 2)*k/r;'

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 118, in codegen.rst

Failed example:

    jscode(e.rhs, assign_to=e.lhs, contract=False)

Expected:

    Dy[i] = (y[i + 1] - y[i])/(t[i + 1] - t[i]);

Got:

    'Dy[i] = (y[i + 1] - y[i])/(t[i + 1] - t[i]);'

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 151, in codegen.rst

Failed example:

    octave_code(f(x) + g(x) + g(mat), user_functions=custom_functions)

Expected:

    existing_octave_fcn(x) + my_fcn(x) + my_mat_fcn([1 x])

Got:

    'existing_octave_fcn(x) + my_fcn(x) + my_mat_fcn([1 x])'

**********************************************************************

File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 190, in codegen.rst

Failed example:

    pprint(expr)

Expected:

                     ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γₛ⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ────────────────────────────────

                     3              

                  rᵢₛ               

Got:

                      ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γ_s⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ─────────────────────────────────

                      3              

                  r_is               

**********************************************************************
```

moorepants commented:
I think you have to use `pprint()` explicitly or the `init_printing()` function. Some of the others just need quotes.

tokencolour committed:
[WIP]Code Generation Documentation

Added pprint and init_printing
Extends #10276
Fixes #8889

tokencolour commented:
I have added pprint and init_printing. Last time I added inverted commas, I got [this](https://travis-ci.org/sympy/sympy/jobs/97627299#L4145). I am really hoping this time the tests pass.

moorepants commented:
Are you running the doctests locally? If you aren't you should be, as the iteration on check if things pass is much faster than waiting for Travis. If you make use of conda you can easily switch among the various Python versions for testing.

tokencolour commented:
I tested all the examples locally before including them in the doc with and without pretty printing. I understand that this is the faster method.

tokencolour commented:
@moorepants I can't guess what's wrong with (this)[https://travis-ci.org/sympy/sympy/jobs/99649081#L4234]. I had no such error when I built the docs on my machine. It rendered as expected. I am making another commit since I have tried to fix other errors. Kindly take a look at this.

tokencolour committed:
Code Generation Documentation

Fixed a pretty printing error
Extends #10276
Fixes #8889

tokencolour commented:
In the latest Travis test, I get the following error on Python 3.2, 3.3, 3.4, 
```
File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 191, in codegen.rst

Failed example:

    pprint(expr)

Expected:

                      ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γ_s⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ─────────────────────────────────

                      3              

                  r_is             

Got:

                     ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γₛ⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ────────────────────────────────

                     3              

                  rᵢₛ               

****************************************
```
In the previous to the latest Travis test, I got the following error, on Python 2.6, 2.7 and pypy
```
File "/home/travis/build/sympy/sympy/doc/src/tutorial/codegen.rst", line 191, in codegen.rst

Failed example:

    pprint(expr)

Expected:

                     ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γₛ⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ────────────────────────────────

                     3              

                  rᵢₛ               

Got:

                      ⎛     2       ⎞

    I_z⋅S_z⋅γᵢ⋅γ_s⋅kᵢ⋅⎝3⋅cos (β) - 1⎠

    ─────────────────────────────────

                      3              

                  r_is               

**********************************************************************
```
What should I do about this? What am I doing wrong?

tokencolour commented:
Also I can't seem to figure out the reason behind [this](https://travis-ci.org/sympy/sympy/jobs/99682162#L4234) error. Everything built fine and rendered correctly on my machine.
@moorepants Please help me.

moorepants committed:
Fixed all of the codegen.rst doctest printing errors for Python 3.5.

moorepants committed:
Cleaned up formatting and removed whitespace.

moorepants committed:
reworded the intro and fixed spelling errors

moorepants commented:
I submitted a PR to your branch with most of the fixes needed.

tokencolour committed:
Merge pull request #2 from moorepants/codegen-doctests

Fixes to the codegen tutorial.

tokencolour commented:
Lessons drawn from your PR
* I'll be careful while adding whitespaces.
* I'll use better formatting of lines.
* I'll use more of cross-referencing in the docs.
* I'll be extra careful in my spellings.
* Those extra lines added in the beginning for each doctest are a blessing.
Thank you, @moorepants , thank you for bearing with a novice like me.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -1,95 +1,105 @@
-       |
-    code printers
+py:func:`sympy.utilities.autowrap` uses codegen, and codegen uses the code
+printers. py:func:`sympy.utilities.autowrap` does everything: it lets you go
```
These links don't seem to work.

tokencolour replied:
Shouldn't this be py:func:`sympy.utilities.autowrap.autowrap`

moorepants replied:
Yes.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -1,95 +1,105 @@
-An example of this is powers in C. x**2 prints as pow(x, 2) instead of x*x. 
-Other optimizations (like mathematical simplifications) should happen 
-before the code printers.
+Currently, :py:func:`sympy.cse` is not applied automatically anywhere in this
```
This link doesn't work either.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -1,95 +1,105 @@
+special cases of the other prints in SymPy (str printer, pretty printer, etc.).
+
+An important distinction is that the code printer has to deal with assignments
+(using the ``sympy.printing.codeprinter.Assignment`` object).This serves as
```
This link doesn't work.

tokencolour replied:
These links do work on my local machine.

moorepants replied:
Didn't work locally for me...

tokencolour replied:
What should I do, scrap the links part?

moorepants replied:
If they work for you, I guess leave them in.

asmeurer replied:
I think to make it a link you need to use :ref:\`sympy.printing.codeprinter.Assignment\` (my rst is rusty so that may not be exactly the right thing)

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -98,27 +108,23 @@ anything. A use case for ``Piecewise``::
-    >>> y = IndexedBase('y', shape=(len_y,))
-    >>> t = IndexedBase('t', shape=(len_y,))
+    >>> y_ = IndexedBase('y', shape=(len_y,))
+    >>> t_ = IndexedBase('t', shape=(len_y,))
```
It'd be nicer if, instead of trailing underscores, we chose non-conflicting variable names.

tokencolour replied:
I'll change the variable names.

moorepants commented on doc/src/tutorial/codegen.rst:
```diff
@@ -127,95 +133,76 @@ looped over::
+    >>> k_i, g_i, g_s, r_is, I_z, S_z = symbols("k_i, gamma_i, gamma_s, r_is, I_z, S_z")
+    >>> expr = k_i * g_i * g_s / (r_is**3)
+    >>> expr = expr * 2 * I_z * S_z * (3 * (cos(beta))**2 - 1) / 2
+    >>> expr
```
This expression prints differently in Python 2 and 3. I'm not sure why, but we need to fix it.

moorepants replied:
I also don't know what some of the subscripts don't use the unicode versions.

tokencolour replied:
What should be done about it?

moorepants replied:
We need to figure out why it doesn't work.

moorepants commented:
We could combine this page with the "Numeric Computation" page. Maybe that would be the best place for it. They are similar pages.

tokencolour commented:
Some parts of the doc are similar to Numeric Computation i.e. those related to speedy backends, but the doc gives an overview on code generation as a whole. Would it be fine if we just extend Numeric Computation, because it will disturb the flow of the document and a lot of topics will seem to be repeated. What do you think, @moorepants?

moorepants commented:
I think we should merge this with Numeric Computation and remove redundant parts. We can house all of this there instead of in the tutorial. I'm not sure if "Numeric Computation" or "Code Generation" is a better title, probably the former.

tokencolour commented:
That's a good idea!

tokencolour committed:
Code generation documentation

Made some cross links
Changed to non-conflicting variable names.
Fixes #8889
Extends #10276

moorepants commented:
I'm not able to reproduce the unicode printing for Python 3.5 manually:

```
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sympy import symbols, init_printing, cos

In [2]: k_i, g_i, g_s, r_is, I_z, S_z = symbols("k_i, gamma_i, gamma_s, r_is, I_z, S_z")

In [3]: init_printing(use_unicode=True)

In [4]: expr = k_i * g_i * g_s / (r_is**3)

In [5]: expr
Out[5]: 
γᵢ⋅γ_s⋅kᵢ
─────────
      3  
  r_is   

In [7]: beta = symbols('beta')

In [8]: expr = expr * 2 * I_z * S_z * (3 * (cos(beta))**2 - 1) / 2

In [9]: expr
Out[9]: 
                  ⎛     2       ⎞
I_z⋅S_z⋅γᵢ⋅γ_s⋅kᵢ⋅⎝3⋅cos (β) - 1⎠
─────────────────────────────────
                  3              
              r_is               
```

moorepants commented:
I take that back, I was using SymPy 0.7.6.1 above and this is master:

```
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sympy import symbols, init_printing, cos

In [2]: k_i, g_i, g_s, r_is, I_z, S_z = symbols("k_i, gamma_i, gamma_s, r_is, I_z, S_z")

In [3]: beta = symbols('beta')

In [4]: init_printing(use_unicode=True)

In [5]: expr = k_i * g_i * g_s / (r_is**3)

In [6]: expr
Out[6]: 
γᵢ⋅γₛ⋅kᵢ
────────
     3  
  rᵢₛ   

In [7]: from sympy import __version__; print(__version__)
0.7.7.dev
```

So there has been a printing change since 0.7.6.1 that doesn't work equally in Py 2 and Py 3.

moorepants commented:
This PR introduced the difference in Python 2 and 3 and XFAILs the unit tests for the changes. So this creates different behavior:

https://github.com/sympy/sympy/pull/9080

You can avoid these subscripts to fix the error here.

tokencolour committed:
Code Generation Documentation

Fixed the example with the complicated expression
Extends #10276
Fixes #8889

tokencolour commented:
I have not yet merged the document with Numeric-Computation. I'll do so shortly.

tokencolour commented:
@moorepants, Do you get [this](https://travis-ci.org/sympy/sympy/jobs/100033502#L4233) error locally?

moorepants commented:
This is likely to do with theano not being installed on travis.

tokencolour commented:
What should I do about it, @moorepants ?

moorepants commented:
Currently building the docs does not require installation of any of the optional dependencies. A decorator can be added to functions with docstrings that conditionally runs them if the optional dependencies are installed. You will need to figure out how to enabled something like this or remove the lines that invoke the optional dependencies. (if that is the error here)

moorepants commented on doc/src/modules/printing.rst:
```diff
@@ -223,6 +223,50 @@ Mathematica code printing
+
+.. module:: sympy.printing.theanocode
+
+.. autodata:: sympy.printing.theanocode.mapping
```
This line isn't that useful to show in the docs because it isn't very readable.

tokencolour replied:
That's true. I blindly followed the status quo.

tokencolour commented:
Something like `@skipIf` of `unittest` library or `@skipif` of `pytest` ? I'll have to find something similar for doctests.

moorepants commented:
Ask on the mailing list.

tokencolour commented:
I'll ask on the mailing list.
Found this [here](http://docs.astropy.org/en/stable/development/testguide.html#skipping-doctests):
```
In the Sphinx documentation, a doctest requirement can be notated with the doctest-requires directive:

.. doctest-requires:: scipy

    >>> import scipy
    >>> scipy.hamming(...)

```

moorepants commented:
That may work if it is a default sphinx function.

tokencolour commented:
However, all I am doing here is importing a dict `mapping` from theanocode.py .  There are no doctests here. Moreover it does not render intelligibly.

moorepants commented:
The mapping thing should be removed. But all of the imports for the autodoc that require optional dependencies may still fail because I think Sphinx actually imports them to retrieve the docstrings.

tokencolour committed:
Code generation Documentation

Fixed few cross links.
Removed theanocode.mapping
Fixes #8889
Extends #10276

tokencolour committed:
Merged Codegen Doc and Numeric-computation

Merged the two docs
Fixes #8889
Extends #10276

tokencolour commented:
@moorepants , Please take a look at this.

tokencolour committed:
Changed index.rst for tutorial

Removed codegen from tutorial
Fixes #8889
Extends #10276

asmeurer commented on doc/src/modules/numeric-computation.rst:
```diff
@@ -1,3 +1,543 @@
+===============
+Code Generation
+===============
```
I would move this to the bottom of the file, or to a new file.

moorepants replied:
I asked @tokencolour to combine his new document into the numeric computation page but to keep the title "Numeric Computation". It is no longer in the tutorial. I'm all ears for any suggestions on how to organize the content though.

tokencolour commented:
Actually, I was very confused while merging the documents. I couldn't change much as far as the order is concerned. I wanted to keep the 2 discussions separate since speedy backends and code generation do have an intersection of topics but are different when one tries to explain them. I have kept the Numeric Computation part in the later part of the doc. (I'll change that, if you suggest so.)  And that's why the doc ended up being named "Code Generation". I wanted to name it "Code Generation and Numeric Computation" and after being in a dilemma for some time I put them differently as equi-headings. IMHO, keeping both the discussions in one doc is a good idea since both are similar and explaining them at one place might avoid any confusions. 

@moorepants , @asmeurer Kindly suggest improvements.
@moorepants Thanks for bearing with me.

asmeurer commented:
My personal preference is to have separate files. There is enough content here for a file of its own, and it's easy enough to link the two documents together, either by cross-referencing or by putting both in a common toc file.

moorepants commented:
As far as I could tell, the only thing covered in the current "Numeric Computation" document that @tokencolour hasn't more thoroughly covered in the new work is subs/evalf. So I don't see why these shouldn't be merged in some way.

Another option would be to create two documents:

Numeric Computation

This would only show the methods of numerically evaluating sympy expressions without any mention of what is going on behind the scenes. So examples of lambdify, autowrap, ufuncify, evalf, etc. (similar to what's there but expanded with all the examples in the new documentation). 

Code Generation

This document would talk about the printers and code generators exclusively along with the explanation of the internals of things like autowrap, but wouldn't have the examples that show the numeric evaluation.

The first document would be for users that need to either evaluate their SymPy expressions numerically or those who need to speed it up and/or vectorize the evaluations.  The second document would be for users that need to control and utilize the code generating levels.

The two documents currently have too much crossover so it is confusing having the same information in two places. Either we need one document that organizes the content in a sane way or we can create two docs that don't have redundancies.

moorepants commented:
@tokencolour This is really close to being done. It'd be great to get it in instead of letting it linger. What do you think about what's been discussed?

tokencolour commented:
@moorepants , I am sorry. I thought it was the collaborators' call. 

>Another option would be to create two documents:

>Numeric Computation

>This would only show the methods of numerically evaluating sympy expressions without any mention of what is going on behind the scenes. So examples of lambdify, autowrap, ufuncify, evalf, etc. (similar to what's there but expanded with all the examples in the new documentation).

>Code Generation

>This document would talk about the printers and code generators exclusively along with the explanation of the internals of things like autowrap, but wouldn't have the examples that show the numeric evaluation.

The ufuncify part of the codegen doc would water down without the examples. It contains a comparison of speeds between the backends. However, the codegen doc does not explicitly explain lambdify, evalf, Theano etc. If we are to keep the docs separately, maybe keeping CodeGen as a whole, and cross-referencing it in Numeric-Computation when we talk about ufuncify, might be fine. Should I continue with this line of thought, and make 2 docs in same toctree? What do you think?

I tried to merge the docs and got confused and the best I could come up with was the last commit.

tokencolour commented:
@moorepants , What do you suggest?

asmeurer commented:
I agree with that plan (that's basically the same thing as what I suggested above).

tokencolour commented:
@moorepants , Do you agree? 
@asmeurer Thank you for your comments.

moorepants commented:
Assuming "the plan" refers to moving back to two documents, that's fine. I just don't want to lose the explanations of how the code printers relate to the higher level functions like autowrap and lambdify.

tokencolour commented:
Thank you @moorepants , I'll come up with a commit shortly. I'll make sure we don't lose those explanations.

tokencolour committed:
Code Generation Documentation

Two separate docs
Fixes #8889
Extends #10276

tokencolour commented:
@moorepants, @asmeurer, Please take a look at this and suggest the required improvements.

moorepants commented on doc/src/modules/codegen.rst:
```diff
@@ -0,0 +1,539 @@
+===============
```
Note that there is a page called 'codegen" already that focused just on the code generation class:

http://docs.sympy.org/latest/modules/utilities/codegen.html

The above page is strictly printing the documentation for that one modue: codegen.

What can we do to not have two pages called the same thing? There should at least be some cross referencing. Is there a better name for this page because it includes things about printing, code generation, and automatic wrapping? This made me realize why I request this content be in the "Numerical Computation" page, because it seems like a better title.

Maybe the bigger issue is that our docs currently mix prose documentation with API reference and there is no clear place for documentation that encompasses information from multiple modules.

moorepants commented on doc/src/modules/codegen.rst:
```diff
@@ -0,0 +1,539 @@
+    I*S*gamma_1*gamma_2*k*(3*Cos[beta]^2 - 1)/r^3
+
+Codegen (sympy.utilities.codegen)
+---------------------------------
```
Needs a cross ref to the page mentioned above.

moorepants commented on doc/src/modules/numeric-computation.rst:
```diff
@@ -1,3 +1,5 @@
+.. _numeric_computation:
```
This page doesn't seem to mention autowrap or all of the functions in the autowrap module.

moorepants commented:
I still don't really like the organization but I'm thinking a bigger documentation cleanup is necessary to do something about it. We can merge this once the tests pass.

tokencolour committed:
Code Generation Documentation

Added more cross refs
Autowrap included in numeric-computation.
Fixes #8889
Extends #10276

tokencolour commented:
@moorepants , Please take a look at this.

tokencolour commented:
I haven't changed the name of the doc. You are right, it should be named differently to avoid confusion. Should we rename the prose one as "Structural Details of Code Generation with Sympy" and let the one with the API reference be called "Code Generation" ? Reorganizing the docs will surely help.

asmeurer commented:
Are any additional changes planned for this?

moorepants commented:
Let's merge this now. I'm happy with it. We can discuss doc reorg at #10456.

moorepants commented:
@tokencolour That name change you suggest is fine.

tokencolour committed:
Changed the name

Fixes #8889
Extends #10276

tokencolour commented:
I've changed the name of the doc. I'll be happy if this gets merged. Thank you, @moorepants and @asmeurer .

moorepants commented:
Thanks @tokencolour!
